File: tools/optimize-images.py
#!/usr/bin/env python3
import os
from PIL import Image
import glob

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
IMAGES_ROOT = os.path.join(ROOT, 'images')

SIZES = [
    ('thumb', 400, 80),
    ('medium', 800, 85),
    ('full', 1600, 90),
]

EXTENSIONS = ('.jpg', '.jpeg', '.png', '.tiff', '.webp', '.heic')

def sanitize_name(name):
    return name.replace(' ', '-')

def process_image(filepath):
    dirpath = os.path.dirname(filepath)
    base = os.path.splitext(os.path.basename(filepath))[0]
    sanitized = sanitize_name(base)

    try:
        with Image.open(filepath) as img:
            original_width, original_height = img.size
            for name, width, quality in SIZES:
                use_width = min(width, original_width) if original_width else width
                # Preserve aspect ratio
                ratio = use_width / float(original_width)
                new_height = int(original_height * ratio)
                resized = img.resize((use_width, new_height), Image.LANCZOS)

                webp_out = os.path.join(dirpath, f"{sanitized}-{name}.webp")
                jpg_out = os.path.join(dirpath, f"{sanitized}-{name}.jpg")

                # Save WebP
                try:
                    resized.save(webp_out, 'WEBP', quality=quality)
                except Exception as e:
                    print(f"Warning: failed to write WebP for {webp_out}: {e}")

                # Save JPEG
                try:
                    # Convert to RGB if image has alpha channel for JPEG
                    if resized.mode in ("RGBA", "LA"):
                        bg = Image.new("RGB", resized.size, (255,255,255))
                        bg.paste(resized, mask=resized.split()[3])
                        bg.save(jpg_out, 'JPEG', quality=quality)
                    else:
                        resized.convert('RGB').save(jpg_out, 'JPEG', quality=quality)
                except Exception as e:
                    print(f"Warning: failed to write JPG for {jpg_out}: {e}")

                print(f"Wrote: {webp_out} and {jpg_out}")

    except Exception as e:
        print(f"Failed to process {filepath}: {e}")


def main():
    files = []
    for ext in EXTENSIONS:
        files.extend(glob.glob(os.path.join(IMAGES_ROOT, '**', f'*{ext}'), recursive=True))

    files = [f for f in files if '/originals/' not in f.replace('\\','/')]
    print(f"Found {len(files)} images to process")
    for f in files:
        process_image(f)

if __name__ == '__main__':
    main()

File: tools/optimize-images.js
#!/usr/bin/env node
const fg = require('fast-glob');
const path = require('path');
const sharp = require('sharp');
const fs = require('fs-extra');

const ROOT = path.resolve(__dirname, '..');
const IMAGES_ROOT = path.join(ROOT, 'images');
const IGNORES = ['**/originals/**'];

const SIZES = [
  { name: 'thumb', width: 600, quality: 80 },
  { name: 'medium', width: 1200, quality: 85 },
  { name: 'full', width: 1920, quality: 90 },
];

const EXTENSIONS = ['jpg', 'jpeg', 'png', 'tiff', 'webp', 'heic'];

function sanitizeName(name) {
  return name.replace(/\s+/g, '-');
}

async function optimizeFile(file) {
  try {
    const fullPath = path.resolve(file);
    const dir = path.dirname(fullPath);
    const ext = path.extname(fullPath).slice(1).toLowerCase();
    const base = path.basename(fullPath, path.extname(fullPath)).replace(/\s+/g, '-');
    if (IGNORES.some(ignore => file.includes(ignore))) return;
    if (!EXTENSIONS.includes(ext)) return;

    const image = sharp(fullPath, { limitInputPixels: false });
    const metadata = await image.metadata();
    const originalWidth = metadata.width || SIZES[SIZES.length - 1].width;

    for (const size of SIZES) {
      const width = Math.min(size.width, originalWidth);
      const webpOut = path.join(dir, `${base}-${size.name}.webp`);
      const jpgOut = path.join(dir, `${base}-${size.name}.jpg`);

      // Create WebP
      await image
        .clone()
        .resize({ width, withoutEnlargement: true })
        .webp({ quality: size.quality })
        .toFile(webpOut);

      // Create JPEG fallback
      await image
        .clone()
        .resize({ width, withoutEnlargement: true })
        .jpeg({ quality: size.quality })
        .toFile(jpgOut);

      console.log(`Wrote: ${webpOut} and ${jpgOut}`);
    }
  } catch (err) {
    console.error('Error optimizing', file, err);
  }
}

async function run() {
  console.log('Scanning images...');
  const patterns = EXTENSIONS.map(ext => `images/**/*.${ext}`);
  const files = await fg(patterns, { dot: false, ignore: IGNORES });

  console.log(`Found ${files.length} images to process`);
  for (const file of files) {
    await optimizeFile(file);
  }
  console.log('Done optimizing images');
}

run().catch(err => {
  console.error(err);
  process.exit(1);
});
